SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -76,18 +76,34 @@ void SpatialVirtualWorkerTranslationManager::ConstructVirtualWorkerMappingFromQu
 			const Worker_ComponentData& Data = Entity.components[j];
 			// System entities which represent workers have a component on them which specifies the SpatialOS worker ID,
 			// which is the string we use to refer to them as a physical worker ID.
-			if (Data.component_id == SpatialConstants::WORKER_COMPONENT_ID)
+			if (Data.component_id == SpatialConstants::SERVER_WORKER_COMPONENT_ID)
 			{
 				const Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data.schema_type);
 
-				const FString& WorkerType = SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::WORKER_TYPE_ID);
+				// The translator should only acknowledge workers that are ready to begin play. This means we can make
+				// guarantees based on where non-GSM-authoritative servers canBeginPlayTrue=true as an AddComponent
+				// or ComponentUpdate op. This affects how startup Actors are treated in a zoned environment.
+				const bool bWorkerIsReadyToBeginPlay = SpatialGDK::GetBoolFromSchema(ComponentObject, SpatialConstants::SERVER_WORKER_READY_TO_BEGIN_PLAY_ID);
+				if (!bWorkerIsReadyToBeginPlay)
+				{
+					continue;
+				}
 
 				// TODO(zoning): Currently, this only works if server workers never die. Once we want to support replacing
 				// workers, this will need to process UnassignWorker before processing AssignWorker.
-				if (WorkerType.Equals(SpatialConstants::DefaultServerWorkerType.ToString()) &&
-					!UnassignedVirtualWorkers.IsEmpty())
+				if (!UnassignedVirtualWorkers.IsEmpty())
 				{
-					AssignWorker(SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::WORKER_ID_ID));
+					// If we didn't find all our server worker entities the first time, future query responses should
+					// ignore workers that we have already assigned a virtual worker ID.
+					const PhysicalWorkerName QueriedWorkerName = SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::SERVER_WORKER_NAME_ID);
+					TArray<FString> ExistingWorkerNames;
+					ExistingWorkerNames.SetNum(VirtualToPhysicalWorkerMapping.Num());
+					VirtualToPhysicalWorkerMapping.GenerateValueArray(ExistingWorkerNames);
+					if (ExistingWorkerNames.Contains(QueriedWorkerName))
+					{
+						continue;
+					}
+					AssignWorker(QueriedWorkerName);
```

Comment:
Might be worth doing the above logic inside `AssignWorker`

Reply:
Changed the logic so it's just an extra case inside the if, I think that should stay where I've put it now, but see what you think?

Reply:
Sorry to reopen, but I'd really prefer the already assigned check be inside AssignWorker, because that's likely where we'll want it when we allow disconnect/connect. I'm happy to have that be a follow-up PR though.